ramips: add support for Dlink DIR-853 A3

kar200 opened the pull request:
Specifications:
* SoC: MT7621AT
* RAM: 256MB
* Flash: 128MB NAND flash
* WiFi: MT7615DN (2.4GHz+5Ghz) with DBDC
* LAN: 5x1000M
* Firmware layout is Uboot with extra 96 bytes in header
* Base PCB is DIR-1360 REV1.0
* LEDs Power Blue+Orange,Wan Blue+Orange,WPS Blue,"2.4G"Blue, "5G" Blue,
 USB Blue
* Buttons Reset,WPS, Wifi


lan      factory 0xe000              (label - same as stock firmware)
wan      factory 0xe006            (same as stock)
2.4 GHz  factory 0x4+bitswap (same as stock)
5.0 GHz  factory 0x4                 (Same as stock)

Flashing instruction:

The Dlink "Emergency Room" cannot be accessed through the reset
button on this device. You can either use console or use the
encrypted factory image availble in the openwrt forum [or here](https://www.mediafire.com/file/z8jmmfz6nwbw1ys/openwrt-21.02.0-rc2-ramips-mt7621-dlink_dir-853-a3-squashfs-factory-enc.bin/file)
The image is 21.02-rc2 encrypted ready to be flashed.

Once the encrypted image is flashed throuh the stock Dlink web
interface, the sysupgrade images can be used.

Header pins needs to be soldered near the WPS and Wifi buttons.

The layout for the pins is (VCC,RX,TX,GND). No need to connect the VCC.

the settings are:

Bps/Par/Bits          : 57600 8N1
Hardware Flow Control : No
Software Flow Control : No

Connect your client computer to LAN1 of the device
Set your client IP address manually to 192.168.0.101 / 255.255.255.0.
Call the recovery page or tftp for the device at http://192.168.0.1
Use the provided emergency web GUI to upload and flash a new firmware to
the device

At the time of writing the the wireless config need to be setup by
editing the wireless config file. A sample wireless config file can
be found on the openwrt forum as well.

Signed-off-by: Karim Dehouche <karimdplay@gmail.com>

kar200 committed:
ramips: add support for Dlink DIR-853 A3

Specifications:
* SoC: MT7621AT
* RAM: 256MB
* Flash: 128MB NAND flash
* WiFi: MT7615DN (2.4GHz+5Ghz) with DBDC
* LAN: 5x1000M
* Firmware layout is Uboot with extra 96 bytes in header
* Base PCB is DIR-1360 REV1.0
* LEDs Power Blue+Orange,Wan Blue+Orange,WPS Blue,"2.4G"Blue, "5G" Blue,
 USB Blue
* Buttons Reset,WPS, Wifi

lan      factory 0xe000      (label - same as stock firmware)
wan      factory 0xe006      (same as stock)
2.4 GHz  factory 0x4+bitswap (same as stock)
5.0 GHz  factory 0x4         (Same as stock)

Flashing instruction:

The Dlink "Emergency Room" cannot be accessed through the reset
button on this device. You can either use console or use the
encrypted factory image availble in the openwrt forum.

Once the encrypted image is flashed throuh the stock Dlink web
interface, the sysupgrade images can be used.

Header pins needs to be soldered near the WPS and Wifi buttons.

The layout for the pins is (VCC,RX,TX,GND). No need to connect the VCC.

the settings are:

Bps/Par/Bits          : 57600 8N1
Hardware Flow Control : No
Software Flow Control : No

Connect your client computer to LAN1 of the device
Set your client IP address manually to 192.168.0.101 / 255.255.255.0.
Call the recovery page or tftp for the device at http://192.168.0.1
Use the provided emergency web GUI to upload and flash a new firmware to
the device

At the time of writing the the wireless config need to be setup by
editing the wireless config file. A sample wireless config file can
be found on the openwrt forum as well.

Signed-off-by: Karim Dehouche <karimdplay@gmail.com>

adschm commented:
So, this device requires soldering to be flashed?

adschm commented on target/linux/ramips/dts/mt7621_dlink_dir-853-a3.dts:
```diff
@@ -0,0 +1,199 @@
+			gpios = <&gpio 14 GPIO_ACTIVE_LOW>;
+		};
+
+		led_net_orange: net_orange{
```
missing space before bracket

kar200 replied:
Ok I will fix that.

adschm commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -275,6 +275,26 @@ define Device/dlink_dir-1960-a1
 TARGET_DEVICES += dlink_dir-1960-a1
 
+define Device/dlink_dir-853-a3
+  $(Device/dsa-migration)
```
It appears you can just use the dlink_dir-xx60-a1 definition here as well, although the name doesn't match.

kar200 replied:
> So, this device requires soldering to be flashed?

Yes it does. The emergency room does not work using the reset button (but works from serial).

kar200 replied:
> It appears you can just use the dlink_dir-xx60-a1 definition here as well, although the name doesn't match.

Yes and I did use that but since they share a common dtsi as well I thought it would be better to treat it as a separate device.

kar200 replied:
I went ahead and made the changes.

adschm commented on target/linux/ramips/mt7621/base-files/etc/board.d/01_leds:
```diff
@@ -29,6 +29,11 @@ gehua,ghl-r-001)
+dlink,dir-853-a3)
+	ucidef_set_led_netdev "wan" "wan" "$boardname:blue:net" "wan"
+	ucidef_set_led_netdev "wlan2g" "WiFi 2.4GHz" "$boardname:blue:wlan2g" "wlan0"
+	ucidef_set_led_netdev "wlan5g" "WiFi 5GHz" "$boardname:blue:wlan5g" "wlan1"
```
Please use a DTS trigger for the WiFi leds.

adschm commented:
Please use your full name for Signed-off-by ...

GalacticLion7 commented:
No merging?

mpratt14 commented:
@kar200 if you deleted only the branch, it can be saved, but if you deleted the fork, I think it's not possible

you can checkout your PR to a new branch and force push it back
However, it looks like the branch is already aligned with the PR

I recommend NOT using github command-line tools. all of it can be done with pure git

```
#add openwrt as a remote alias
git remote add openwrt https://github.com/openwrt/openwrt.git

#fetch all PRs
git fetch openwrt +refs/pull/*:refs/remotes/openwrt/pull/*

#checkout your PR
git checkout openwrt/pull/3318/head

#delete your master branch and checkout a new one
git branch -D master
git checkout -b master

#force push back to github
git push --force origin master
```

if the branch was the only problem, you should be able to reopen PR at the bottom

if that doesn't work contact github support
otherwise, this PR is screwed

also some advice for the future:

NEVER delete your fork
NEVER close your own PR unless you have given up forever on the subject
and don't use master branch for PRs, learn how to branch to stage your changes, and keep master branch up to date with openwrt/master without any other commits

kar200 commented:
> NEVER delete your fork
> NEVER close your own PR unless you have given up forever on the subject
> and don't use master branch for PRs, learn how to branch to stage your changes, and keep master branch up to date with openwrt/master without any other commits

Thanks for the tip I wish I knew before. I am still learning and the delete habbit comes form the fact I always manage to mess the master tree everytime I try something and end up not being able to recover - hence just easier to delete and start again. I guess I did not know I would need it to reopen the PR.

The PR was closed mainly cause there was issues with the device booting where it would lock up if both wireless cards were set to active (DBDC issue) which is not comletely fixed but now with manual options in the wireless config file it works. 

And now I have a small user base that are using the router with openwrt and happy with it. Rather than keep doing it through the forum I thought it is the right time to have it officially supported (Bonus if it can make it for 21.02).

As you said it is the fork that is deleted and not just the branch. And after the heated debate around a duplicate PR I will just keep trying and send and email to Github support, they will probably won't understand why there are around 200 deleted forks from the same repo :)

Anyway thank you for your time and explanation (and tips).

Have a good evening

mpratt14 commented:
@kar200 also an alternative to github PR is to send your patch using git send-email

this requires exporting your commits in patch file format
`git format-patch -<n> <sha>`
where you replace `<sha>` is the hash of the commit
and replace `<n>` with the number of commits

you can open the patch file like any other text document to make sure its ok

then setup SMTP between your computer and your mail service
https://git-send-email.io/

and send an email using the command line to the mailing list with your patches as an attachment

https://openwrt.org/docs/guide-developer/working-with-git-email

kar200 commented:
@mpratt14 Thanks I will have a look and try that (hope it doesnt end up in a new request lol).

I will let you know if it works.

kar200 commented:
Thanks again @mpratt14 for the time and patience with this. I have contacted github support and they restored my tree. 

Since this is my master tree on github and I don't want to mess it up, is it possible for me to edit directly for this page?

kar200 commented:
For the reviewers: I have not modified anything yet on this pull request so please bear with me until it is done. 

The changes won't be that many and there is a full thread on the forum about people already using the device successfully.

https://forum.openwrt.org/t/dlink-dir-853-a3/70548/226

mpratt14 commented:
@kar200 glad you were able to save your PR

I do not know what you mean by "edit directly for this page" so I will explain in general:

always note the status right under the title
" Open  kar200 wants to merge 5 commits into openwrt:master from kar200:master"

the PR is tracking the source branch (in this case, kar200:master)
and comparing it against the base branch (openwrt:master)
and the difference in commits between the branches is 5 right now

as long as your PR is open, it will update based on the source branch that it is tracking
in other words, as long as it is open, any changes to your master branch will take effect here as well

so all you need to do now is make whatever changes you need to do on your local git branch
and then force push it back to github
`git push --force origin master`

if you need help combining your 5 commits into 1 commit feel free to ask

kar200 commented:
> @kar200 glad you were able to save your PR
> 
> I do not know what you mean by "edit directly for this page" so I will explain in general:


My bad I realised I was talking rubbish about updating here since there are merge conflicts anyway. What I meant was what I did the first time I sent this I just made the small modifications directly from the web interface to avoid any mistakes (which is obviously the incorrect way to go about it). 

I am reading about it at the moment.

> the PR is tracking the source branch (in this case, kar200:master)
> and comparing it against the base branch (openwrt:master)
> and the difference in commits between the branches is 5 right now

That explains a bit better. I thought once the pull request is sent from the fork that's it. A new one needs to be sent again wit the changes and then the magic happens! I did not know the changes are constantly tracked and updated from the fork to the master.

Since you are offering nicely how do you combine everything into a new single commit lol? If I fix all the merge conflicts in my own fork and change everything to a single commit and the force push to my fork would this pull request automatically pick up everything and change it accordingly?

mpratt14 commented:
first handle putting the 5 commits into 1
correct the merge conflict later...

a simple way to roll back commits is with git reset
this is ideal for rolling back 1 commit, or a small number
`HEAD` is an alias for the hash of the very top commit of the branch
`HEAD^` represents the parent commit of HEAD (the commit right before HEAD)
a "soft" reset moves the changes from commit back to staging, so that it is ready to be committed again
a "hard" reset removes the changes completely

so you would first do this
`git reset --soft HEAD^`

then see your status, that the changes are staged (green)
`git status`

verify that the top commit is now gone (q to exit)
`git log`
or
`git log --oneline`

take a look at the top commit if you like (q to exit)
`git show HEAD`
or the commit before HEAD
`git show HEAD^` or `git show HEAD~1`
or 2 commits before HEAD
`git show HEAD~2`
... ...

repeat the soft reset a few times, then when the top commit is your first one, add the changes to HEAD
`git commit --amend`

and whenever you want to add changes to the current commit, you do `--amend` instead of `--signoff`
to merge the changes to the current commit instead of creating a new one

also note that your commit title is very long for the first one
you need an extra empty line between the title and the rest

when there is no changes staged under `git status`
you can do `git commit --amend` at any time to edit the commit message

finally push it back
`git push --force origin master`

kar200 commented:
Thanks @mpratt14 that worked a treat. By the way once my commit is ready do I push back at the same state it is now (a couple of thousnads commits behind + merge conflicts) or do I solve the conflicts first on my PC and then force push?

mpratt14 commented:
@kar200 ok good job, now for merge conflict

the conflict occurs between this commit and commits that are not yet in your branch (because they are new changes in openwrt master)

the first step is to perform a pull during a rebase

a rebase is when you temporarily remove a commit to make changes to a commit that is not on top
after rebase, the commits that were temporarily removed get put back on top

in this case we want to pull openwrt master branch during the rebase

you can do git pull during a rebase state (after git rebase command)
however, git pull also has a shortcut option for doing both things at the same time

`git pull --rebase https://github.com/openwrt/openwrt.git master`

you can further simplify this by adding a remote alias if you didn't already
`git remote add openwrt https://github.com/openwrt/openwrt.git`
then it would be
`git pull --rebase openwrt master`

after that command, it will print info about the conflict and then exit
it's up to you to fix the conflict before finishing the rebase with `git rebase --continue`

instead of reading whats printed during rebase, its better to see whats going on with `git status`
you would see 2 files labeled as "both modified"

during a conflict, git puts markers in the file that indicate the difference between what you are pulling vs what you are rebasing
something like this

```
<<<<<< HEAD
some lines
that exist
in openwrt master
========
some lines
that exist
in your commit
>>>>>>> (hash) ramips: your commit title
```

so for each file labeled "both modified" you open them in the editor, search for "==="
and then edit the file in a way that accommodates both new changes from master and your changes
usually this means simply removing the conflict markers and adding empty lines where it makes sense

if you get confused you can do
`git rebase --abort`
and try again

when done add the files back to staging
`git add -A`
and finish the rebase
`git rebase --continue`

and then push it to github
`git push --force origin master`

after that your branch and PR will be up-to-date
(currently it is like 2700 commits behind)

kar200 commented:
@mpratt14 good job to you lol. I just followed the instructions.This was a very good "free" lecture about the important git commands. I used to delete the branches because I could not rebase correctly.

All done now thanks again. I guess with the position in the queue this will take some time to get merged (if there are no issues).

For the reviewers, this has been tested already (although built by me) on the openwrt forum. There are 2 commits in the pipeline (one to have the correct dlink header and the other one to have encryption) where if they get merged and work correctly the images of nearly all the recent dlink devices can be flashed directly from the web interface.

Lucky1openwrt commented on target/linux/ramips/dts/mt7621_dlink_dir-853-a3.dts:
```diff
@@ -0,0 +1,206 @@
+};
+
+&pcie0 {
+	mt76@0,0 {
```
this has been changed at some point from `mt76@0,0` to `wifi0: wifi `

kar200 replied:
Thanks, I will test it and update this.

mpratt14 replied:
label is not needed if you don't use it
it honestly should be

`wifi@0,0,0`

and it does not require testing, it's just naming convention

kar200 replied:
Actually it did work with it so I fill just use the same for consistency

Lucky1openwrt commented on target/linux/ramips/dts/mt7621_dlink_dir-853-a3.dts:
```diff
@@ -0,0 +1,206 @@
+
+		partition@4980000 {
+			label = "firmware2";
+			compatible = "openwrt,uimage", "denx,uimage";
```
you don't need to define anything about firmware2 as it's not used
it just creates extra rootfs & ubi partitions that are not used

kar200 replied:
I know but since the device uses the same partition layoput as all the other dlink Nand devices I just ussed the same layout (I thought maybe there was some conversations and then agreement about using this). I wll change this as well.

Lucky1openwrt replied:
I'm hoping to fix the other link devices if and when a pull request happens

mpratt14 replied:
is there 2 kernels and 2 rootfs from the factory?

if so, it is a candidate for dual-boot maybe?

if not, the space can still be used with the mtd-concat driver

Lucky1openwrt replied:
the boot loader backs up the kernel from firmware
but you could recover some room tho

mpratt14 commented on target/linux/ramips/dts/mt7621_dlink_dir-853-a3.dts:
```diff
@@ -0,0 +1,206 @@
+		#size-cells = <1>;
+
+		partition@0 {
+			label = "Bootloader";
```
it looks like uboot so make the label

`"u-boot"`

kar200 replied:
It is indeed. I will change it 
PS: I just followed the other dlink devices that use the same partition scheme. But I will update this.

mpratt14 commented on target/linux/ramips/dts/mt7621_dlink_dir-853-a3.dts:
```diff
@@ -0,0 +1,206 @@
+		led-failsafe = &led_power_blue;
+		led-running = &led_power_blue;
+		led-upgrade = &led_net_orange;
+
```
remove extra line

kar200 replied:
Will do.

mpratt14 commented on target/linux/ramips/dts/mt7621_dlink_dir-853-a3.dts:
```diff
@@ -0,0 +1,206 @@
+			linux,default-trigger = "usbport";
+		};
+
+		wlan2g {
+			label = "dir-853-a3:blue:wlan2g";
+			gpios = <&gpio 4 GPIO_ACTIVE_LOW>;
+			linux,default-trigger = "phy0radio";
+		};
+
+		wlan5g {
+			label = "dir-853-a3:blue:wlan5g";
+			gpios = <&gpio 3 GPIO_ACTIVE_LOW>;
+			linux,default-trigger = "phy1radio";
+		};
```
usually we use this

```
linux,default-trigger = "phy0tpt";
linux,default-trigger = "phy1tpt";
```

kar200 replied:
Unfortunately it does not work with the phy option. Only the 2.4ghz work and the 5 stays off all the time.
It seems with these ones it works for both cards. I guess this is due to DBDC maybe?

mpratt14 commented:
to use the mtd-concat driver here:

change firmware partition to
```
		fwconcat0: partition@180000 {
			label = "fwconcat0";
			reg = <0x180000 0x2800000>;
		};
```

change firmware2 partition to:
```
		fwconcat1: partition@4980000 {
			label = "fwconcat1";
			reg = <0x4980000 0x2800000>;
		};
```

add this node to the root node (right under leds)
```
	virtual_flash {
		compatible = "mtd-concat";

		devices = <&fwconcat0 &fwconcat1>;

		partitions {
			compatible = "fixed-partitions";
			#address-cells = <1>;
			#size-cells = <1>;

			partition@0 {
				compatible = "openwrt,uimage", "denx,uimage";
				openwrt,padding = <96>;
				label = "firmware";
				reg = <0x0 0x0>;
			};
		};
	};
```

kar200 commented:
> to use the mtd-concat driver here:
> 
> change firmware partition to
> 
> ```
> 		fwconcat0: partition@180000 {
> 			label = "fwconcat0";
> 			reg = <0x180000 0x2800000>;
> 		};
> ```
> 
> change firmware2 partition to:
> 
> ```
> 		fwconcat1: partition@4980000 {
> 			label = "fwconcat1";
> 			reg = <0x4980000 0x2800000>;
> 		};
> ```
> 
> add this node to the root node (right under leds)
> 
> ```
> 	virtual_flash {
> 		compatible = "mtd-concat";
> 
> 		devices = <&fwconcat0 &fwconcat1>;
> 
> 		partitions {
> 			compatible = "fixed-partitions";
> 			#address-cells = <1>;
> 			#size-cells = <1>;
> 
> 			partition@0 {
> 				compatible = "openwrt,uimage", "denx,uimage";
> 				openwrt,padding = <96>;
> 				label = "firmware";
> 				reg = <0x0 0x0>;
> 			};
> 		};
> 	};
> ```

Thanks, but unfortunately it won't be possible. The second firmware partition gets overwritten for any change
U-boot checks for both firmware partitions if they are "valid" and will overwrite the other one with the valid one unless they are both valid but different it will write the second one with the first one. 

I guess one way to do it would be since the kernel size is set at 4mb and it only checks the kernel (uses the length from the kernel header) we could potentially concat from 4mb offset.

Here is an example of a device update.
````
Check image validation:hdr1_addr[bc180000]hdr2_addr[c0980000]
Image1 Header Magic Number --> OK
Image2 Header Magic Number --> OK
Image1 Header Checksum --> OK
Image2 Header Checksum --> OK
Image1 Data Checksum --> .......................................OK
Image2 Data Checksum --> .......................................OK
Image2 is old image, will update Image2

Image1: OK Image2: Broken

Image2 is borken. Copy Image1 to Image2.
Copy Image:
Image1(0x180000) to Image2(0x4980000), size=0x261238
.......................................ranand_erase: start:4980000, len:20000 
..ranand_erase: start:49a0000, len:20000 
..ranand_erase: start:49c0000, len:20000 
..ranand_erase: start:49e0000, len:20000 
..ranand_erase: start:4a00000, len:20000 
..ranand_erase: start:4a20000, len:20000 
..ranand_erase: start:4a40000, len:20000 
..ranand_erase: start:4a60000, len:20000 
..ranand_erase: start:4a80000, len:20000 
..ranand_erase: start:4aa0000, len:20000 
..ranand_erase: start:4ac0000, len:20000 
..ranand_erase: start:4ae0000, len:20000 
..ranand_erase: start:4b00000, len:20000 
..ranand_erase: start:4b20000, len:20000 
..ranand_erase: start:4b40000, len:20000 
..ranand_erase: start:4b60000, len:20000 
..ranand_erase: start:4b80000, len:20000 
..ranand_erase: start:4ba0000, len:20000 
..ranand_erase: start:4bc0000, len:20000 
....ranand_erase: start:4be0000, len:20000 
.(5192)offs=79560704 piece=0 piece_size=4664 rc=0
Done!

=================================================

`````

mpratt14 commented:
interesting....

can you print uboot-environment?

`printenv`

if any variables control which one is booted, it would be a good candidate for dual-booting

kar200 commented:
> interesting....
> 
> can you print uboot-environment?
> 
> `printenv`
> 
> if any variables control which one is booted, it would be a good candidate for dual-booting


Sure

````
U-Boot 1.1.3 (Dec 20 2018 - 11:06:07)
MT7621 # printenv
bootcmd=tftp
bootdelay=1
baudrate=57600
ethaddr="00:AA:BB:CC:DD:10"
ipaddr=192.168.0.1
serverip=192.168.0.101
stdin=serial
stdout=serial
stderr=serial
BootType=3

Environment size: 162/4092 bytes
````

Lucky1openwrt commented:
this is form my version of the "mt7621_dlink_dir-xx60-a1.dtsi"
`&nand {
	status = "okay";

	partitions {
		compatible = "fixed-partitions";
		#address-cells = <1>;
		#size-cells = <1>;

		partition@0 {
			label = "Bootloader";
			reg = <0x0 0x80000>;
			read-only;
		};

		partition@80000 {
			label = "config";
			reg = <0x80000 0x80000>;
			read-only;
		};

		factory: partition@100000 {
			label = "factory";
			reg = <0x100000 0x40000>;
			read-only;
		};

		partition@140000 {
			label = "config2";
			reg = <0x140000 0x40000>;
			read-only;
		};

		partition@180000 {
			label = "kernel";
			compatible = "openwrt,uimage", "denx,uimage";
			openwrt,padding = <96>;
			reg = <0x180000 0x400000>;
		};

		ubiconcat0: partition@580000 {
			label = "ubiconcat0";
			reg = <0x580000 0x4400000>;
		};

		partition@4980000 {
			label = "kernel2";
			reg = <0x4980000 0x400000>;
		};

		ubiconcat1: partition@4D80000 {
			label = "ubiconcat1";
			reg = <0x4D80000 0x2400000>;
		};

		partition@7180000 {
			label = "mydlink";
			reg = <0x7180000 0x600000>;
			read-only;
		};

		ubiconcat2: partition@7780000 {
			label = "ubiconcat2";
			reg = <0x7780000 0x860000>;
		};
	};
};`

tho concat fails if the end address is 0x80000 so it's a page less
need to make sure your firmware has not used extra partitions allocated tho

mpratt14 commented:
if we could figure out what `BootType` variable does that might help

but be careful because it might have something to do with flashing the bootloader, always have a backup of flash

Lucky1openwrt commented:
boot type 3 is 3 outoff the uboot menu defaut
it's boot from flash

@kar200 note if you change ubi size sysupgrade won't work
you need to flash with factory image
i would love to know if there is a way to get sysupgrade to work when changing ubi size

Lucky1openwrt commented:
when uboot stats it check the 1st partition "kernel" and checks that the uboot image is valid with it's size & checksum
if it differs and is valid it copys to backup
if invalid & backup is valid it restores it 
up to the size in the uboot header so only the kernel

mpratt14 commented:
just throwing another idea out there

one could make a custom sysupgrade script that plants a fake uimage header and fake kernel in the other firmware partition

then it would be possible to use mtd-concat and only be wasting 1 block

to use the extra space one would sysupgrade after initial factory.bin flash

Lucky1openwrt commented:
it is nice to know if you kernel partition fails it's checksum it will get repaired it
I'm not sure if they ever do fail after the fact but it will do it's job and fix it
with the config I posted you end  up with 97M out of 128M 
that's if my memory is correct
default is not enough to install samba v4

kar200 commented:
I wll try with the concat example and see how it goes and let you know. Unfortunately I cannot update the pull request at the moment as there are a few users already using the same partition layout on the forums and they will probably want to update to this if it gets merged.

Lucky1openwrt commented:
if you flash the factory file it's fine as it's replaces the ubi partition

kar200 commented:
@Lucky1openwrt It worked. But I could not flash the factory partition though through Luci - complains about invalid tar image when using the force option.

````
root@OpenWrt:/# df -h
Filesystem                Size      Used Available Use% Mounted on
/dev/root                 4.0M      4.0M         0 100% /rom
tmpfs                   122.9M    244.0K    122.6M   0% /tmp
/dev/ubi0_1              92.7M     52.0K     88.0M   0% /overlay
overlayfs:/overlay       92.7M     52.0K     88.0M   0% /
tmpfs                   512.0K         0    512.0K   0% /dev
root@OpenWrt:/# cat /proc/mtd
dev:    size   erasesize  name
mtd0: 00080000 00020000 "u-boot"
mtd1: 00080000 00020000 "config"
mtd2: 00040000 00020000 "factory"
mtd3: 00040000 00020000 "config2"
mtd4: 00400000 00020000 "kernel"
mtd5: 04400000 00020000 "ubiconcat0"
mtd6: 00400000 00020000 "kernel2"
mtd7: 02400000 00020000 "ubiconcat1"
mtd8: 00600000 00020000 "mydlink"
mtd9: 00860000 00020000 "ubiconcat2"
mtd10: 07060000 00020000 "ubi"
````

This is good to get maximum size. I ll probably not implement this separately for this device and just wait for it to be approved for all the others first.

Lucky1openwrt commented:
can only do factory image via recover / serial or OEM if encrypted
I'm not sure it will make the others'
there are at lest 4 devices off that one DTSI
& will have to make sure all of the partitions are not used on any of the models 
and someone has to do a PR for it

mpratt14 commented:
doing it for this board probably makes it more likely it will be applied to the rest later

> But I could not flash the factory partition though through Luci

the factory.bin is a tar image, and sysupgrade is normal?
or you actually mean the partition?

kar200 commented:
@mpratt14 I agree but the biggest issue with any change of this nature is that this device does not have any recovery method. Once it's bricked you cannot recover unless you open it and solder the pins...etc. I am not sure what dlink did but I am guessing they just didn't define the reset button to the correct gpio in u-boot or something like that since the recovery works from selecting the correct option in console mode.



> the factory.bin is a tar image, and sysupgrade is normal?
> or you actually mean the partition?

I thought it would work by using the web interface and tried that. Obviously sysupgrade expects a tar image and I was trying the factory one so it failed complaining that the tar is invalid - which is correct.

adschm commented on target/linux/ramips/dts/mt7621_dlink_dir-853-a3.dts:
```diff
@@ -0,0 +1,203 @@
+		compatible = "gpio-leds";
+
+		led_power_orange: power_orange {
+			label = "dir-853-a3:orange:power";
```
remove vendor prefix, just orange:power and so on ...

kar200 replied:
Ok. I will (once all the requested changes are addressed so I will have a single new commit with all the changes).

adschm commented:
> 5.0 GHz  factory 0xe000+1 (different from stock - created by DBDC)

What would it be on stock then? Could you possibly add the last byte of the address for each interface, so we could see the increments compared to each other ...

adschm commented on target/linux/ramips/dts/mt7621_dlink_dir-853-a3.dts:
```diff
@@ -0,0 +1,203 @@
+
+&state_default {
+	gpio {
+		groups = "i2c", "uart3", "jtag", "wdt";
```
uart2 missing.

kar200 replied:
Ok

adschm commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -342,6 +342,13 @@ define Device/dlink_dir-2660-a1
 TARGET_DEVICES += dlink_dir-2660-a1
 
+define Device/dlink_dir-853-a3
+  $(Device/dlink_dir-xx60-a1)
```
We should find a better name for this common definition ...

kar200 replied:
Maybe something like Dlink-nand? It seems all the dlink routers on the mt7621 with nand flash share the same partition layout. If possible maybe keep the dtsi the same and rename in the mt7621.mk

kar200 commented:
> > 5.0 GHz  factory 0xe000+1 (different from stock - created by DBDC)
> 
> What would it be on stock then? Could you possibly add the last byte of the address for each interface, so we could see the increments compared to each other ...

I came to notice that for some reason with the latest master (as of today) the 2.4 and 5 interface have the same mac address. They used to be different.

Anyway he is an output of the ip link command from stock and openwrt. I thouhg the "ra" interface was the 2.4 and "rax" is the 2.5? When using a scan it seems the 5ghz interface is the one getting the similar address in stock dlink formware and the 2.4 something different.

````
Label: F48CEBA8F065

Openwrt master:
eth0: f4:8c:eb:a8:f0:65
lan4@eth0: f4:8c:eb:a8:f0:65 
lan3@eth0: f4:8c:eb:a8:f0:65 
lan2@eth0: f4:8c:eb:a8:f0:65
lan1@eth0: f4:8c:eb:a8:f0:65 
wan@eth0: f4:8c:eb:a8:f0:68 
br-lan: f4:8c:eb:a8:f0:65 
wlan0: f4:8c:eb:a8:f0:66 
wlan1: f4:8c:eb:a8:f0:66


Stock Dlink - I removed a few interfaces as there was a few bridges and and other apcli interfaces. 
eth2: f4:8c:eb:a8:f0:65
eth2.1@eth2: f4:8c:eb:a8:f0:65 
eth2.2@eth2: f4:8c:eb:a8:f0:65
eth2.3@eth2:  f4:8c:eb:a8:f0:65
eth2.4@eth2: f4:8c:eb:a8:f0:65
eth3: f4:8c:eb:a8:f0:68
br0:  f4:8c:eb:a8:f0:65
ra0:  f4:8c:eb:a8:f0:66
rax0: f6:8c:eb:c8:f0:66

Wifi scan from my computer (nmcli device wifi list)
F6:8C:EB:C8:F0:66  dlink-F065
F4:8C:EB:A8:F0:66  dlink-F065-5GHz

````

kar200 commented:
I have pushed latest changes. I should have addressed all the changes.

* Removed the vendor prefix
* Added the uart2
* Used the bitflip from the DIR-853-R1 but I have to change 2 bits to get the same as stock (set and unset). Now all the mac addresses match.

````
Openwrt

F6:8C:EB:C8:F0:66  test2      
F4:8C:EB:A8:F0:66  test5   

Stock
F6:8C:EB:C8:F0:66  dlink-F065
F4:8C:EB:A8:F0:66  dlink-F065-5GHz
`````

adschm commented:
I've added a version with minor changes here:
https://git.openwrt.org/?p=openwrt/staging/adrian.git;a=shortlog;h=refs/heads/dir853a1

I've also removed the comment

>    At the time of writing the the wireless config need to be setup by
>    editing the wireless config file. A sample wireless config file can
>    be found on the openwrt forum as well.

Or is this still relevant?

I still do wonder whether this whole set/unset thing is correct for the MAC address. Is there any chance we could get the MAC address assignment from one or a few additional devices? (Just from OEM, no need to flash anything)

Probably it's really just `macaddr_add $basemac 0x200000` ...

kar200 commented:
> Or is this still relevant?

Not really but a few options need to be set. Since the MT76 gets updated often I am not sure. At the moment it works by setting the country to any country rather than leaving it blank otherwise it looses all connectivity at boot time (can be reset though with the reset button).


> I still do wonder whether this whole set/unset thing is correct for the MAC address. Is there any chance we could get the MAC address assignment from one or a few additional devices? (Just from OEM, no need to flash anything)

I will try and get a couple. Any chance I can send them to you directly and not post them here?

If I leave it as it is and do not set anything it will still have the same mac addresses for 2.4 and 5 except the first 2 bytes. I don't know but it should be enough?

kar200 commented:
This is from using my phone as a scanner. I am pretty sure they are from the same router. I will update this when I get more.
Obviously the XX will mean it is the same byte

````
X2:XX:XX:48:XX:XX 
X0:XX:XX:68:XX:XX

X4:XX:XX:6A:XX:XX
X6:XX:XX:4A:XX:XX
````

adschm commented:
Thanks, which is which (2g vs. 5g)? Just edit your message above ...

adschm commented:
Okay, so we get a mixture of +0x20 and -0x20 here. So, I will take the same road here as for the r1 and just take the lan address with local bit instead of guessing ...

> Not really but a few options need to be set. Since the MT76 gets updated often I am not sure. At the moment it works by setting the country to any country rather than leaving it blank otherwise it looses all connectivity at boot time (can be reset though with the reset button).

Then please post an explicit explanation what to do exactly here, so I can add it to the commit message.

kar200 commented:
> Okay, so we get a mixture of +0x20 and -0x20 here. So, I will take the same road here as for the r1 and just take the lan address with local bit instead of guessing ...

Sure. I guess as long as the Mac addresses are based on factory and only a single HEX is changed it should be fine (since they don't travel past the first router)

> Then please post an explicit explanation what to do exactly here, so I can add it to the commit message.
Here it is:

I have tested with the latest changes as of today. Setting the country code is mandatory otherwise the router looses connectivity at the next reboot. This is mandatory and can be done from luci. After setting the country code the router boots correctly. A reset with the reset button will fix the issue and the user has to reconfigure.

Second Issue: This is minor since the router the 5g interface does not come up online although it is not set as disabled. 2 options here. 

1- Either run the "wifi" command. Can be added from LUCI in system - startup - local startup and just add wifi above "exit 0" 

2- Or add the serialize option in the wireless config file as shown below. This one would work and bring both interfaces automatically at every boot.

By the way I did not change anything to the commit nor to the code since I am guessing you will close this pull request and add it from your staging tree. Just letting you know since there is a merge conflict (after you added the DIR-853 R1).

Also it seems that in the leds file the DIR-853-r1 is not in the correct line (just cosmetic). I believe it should be just above the dir-860l)

````
config wifi-device 'radio0'
        option serialize '1'


config wifi-device 'radio1'
        option serialize '1'
````

adschm commented:
Thanks, I will merge the device including the recommended adjustments from the most recent post during the weekend.

adschm commented:
I just found the mistake that probably was the reason for requiring mtd-mac-address in the DTS.
We were not using line break markers correctly in 10_fix_wifi_mac.

I've uploaded a fixed version here:
https://git.openwrt.org/?p=openwrt/staging/adrian.git;a=shortlog;h=refs/heads/dir853fix

This also includes all other changes/fixes, so you might have a look at those as well. Please test this and report whether both Wifi addresses are correct now. If 5g is still wrong, I'd add back the mtd-mac-address again, but probably it's fine without.

kar200 commented:
> This also includes all other changes/fixes, so you might have a look at those as well. Please test this and report whether both Wifi addresses are correct now. If 5g is still wrong, I'd add back the mtd-mac-address again, but probably it's fine without.

Sure I will try that and let you know.

kar200 commented:
These are now the 2 mac addresses with your changes.

````
wlan0: F6:8C:EB:A8:F0:65
wlan1: 76:8C:EB:A8:F0:66
````

And from previous post

````
Openwrt
F6:8C:EB:C8:F0:66  test2      
F4:8C:EB:A8:F0:66  test5   

Stock
F6:8C:EB:C8:F0:66  dlink-F065
F4:8C:EB:A8:F0:66  dlink-F065-5GHz
````

adschm commented:
Thanks. Okay, the 2.4g is as expected after the change (lan address with local bit), but the 5g address still deviates from 0x4 in the first byte.

So, we should keep the mtd-mac-address in DTS.
